statemodify/xdd.py

```diff
@@ -0,0 +1,167 @@
+import itertools
+from glob import glob
+from pathlib import Path
+from typing import Union, List, Type
+
+import numpy as np
+import pandas as pd
+from joblib import Parallel, delayed
+from tqdm import tqdm
+
+
+class XddConverter:
+    """Convert .xdd files to columnar .parquet files"""
+
+    def __init__(
+        self,
+        *,
+        output_path: Union[str, Path] = "./output",
+        allow_overwrite: bool = False,
+        xdd_files: Union[str, Path, List[Union[str, Path]]] = "**/*.xdd",
+        id_subset: Union[None, List[str]] = None,
+        parallel_jobs: int = 4,
+    ):
+        """
+
+        :param output_path:         Path to a folder where outputs should be written; default "./output"
+        :type output_path:          str
+
+        :param allow_overwrite:     If False, abort if files already exist in the output_path; default False
+        :type allow_overwrite:      bool
+
+        :param xdd_files:           File(s) or glob(s) to the .xdd files to convert; default "**/*.xdd"
+        :type xdd_files:            List[str]
+
+        :param id_subset:           List of structure IDs to convert, or None for all; default None
+        :type id_subset:            List[str]
+
+        :param parallel_jobs:       How many files to process in parallel; default 4
+        :type parallel_jobs:        int
+
+        """
+
+        # where to write the parquet files
+        self.output_path = Path(output_path)
+        if not self.output_path.exists():
+            self.output_path.mkdir(parents=True, exist_ok=True)
+        if not self.output_path.is_dir():
+            raise IOError(f"output_path '{self.output_path}' is not a directory")
+        if not allow_overwrite:
+            if len(list(self.output_path.glob("*.parquet"))) > 0:
+                raise IOError(f"parquet files already exist in {self.output_path} but allow_overwrite is False")
+        else:
+            for f in self.output_path.glob("*.parquet"):
+                f.unlink()
+
+        # which xdd files to convert
+        if isinstance(xdd_files, list):
+            self.xdd_files = [Path(g) for sublist in [glob(p) for p in xdd_files] for g in sublist]
+        else:
+            self.xdd_files = [Path(g) for g in glob(xdd_files, recursive=True)]
+        if len(self.xdd_files) == 0:
+            raise IOError(f"no .xdd files found in {xdd_files}")
+
+        # filter to a subset of structure IDs (None means keep all)
+        self.id_subset = id_subset
+
+        # run this many jobs in parallel
+        self.parallel_jobs = parallel_jobs
+
+        # field names for each .xdd entry
+        self.fields = [
```

Comment:
@crvernon and I discussed a bit, and decided to leave as is for now, but look for ways to autodetect the appropriate fields/sizes if possible in the future (or otherwise allow a user to override the default settings).